src/os/bluestore/BlueStore.cc

```diff
@@ -5466,10 +5514,13 @@ int BlueStore::_do_write(
     if (offset >= bp->first &&
 	offset > tail_start &&
 	offset + length >= o->onode.size &&
+	o->tail_offset == tail_start &&
 	o->tail_bl.length() &&
 	(offset / block_size == (o->onode.size - 1) / block_size)) {
       dout(20) << __func__ << " using cached tail" << dendl;
       assert((offset & block_mask) == (o->onode.size & block_mask));
+      // wait for any related wal writes to commit
+      txc->osr->wait_for_wal_on_seq(o->tail_txc_seq);
```

Comment:
This problem only comes up because we are making use of the cached state, 
which doesn't happen during wal replay. The problem is we may have a wal 
op that wrote the data we're caching, but it hasn't committed yet.. and 
then we use the cached data to queue a new write that doesn't need wal and 
orders before it.